### PR TITLE
REF: make pytables get_atom_data non-stateful

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @WillAyd how do we annotate "this returns a subclass of Col"?  Here i've used `Type["Col"]` but afaik that doesnt include subclasses

### CODE REVIEW

**Comment by WillAyd** [comment]
I believed Type is covariant so should work for subclasses. Were you seeing an error?

**Comment by jbrockmendel** [comment]
didnt see an error, was just curious

**Comment by WillAyd** [comment]
Yea Type is covariant by default so works for subclasses. TypeVar is invariant so worth knowing that distinction

**Comment by jbrockmendel** [comment]
updated to remove a comment explained the answer to.

also made unrelated changes to one comment typo and to re-raise on AssertionError

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -2453,19 +2456,22 @@ def set_atom_categorical(self, block):
...
         values = block.values
         codes = values.codes
-        self.kind = "integer"
+
         self.dtype = codes.dtype.name
         if values.ndim > 1:
             raise NotImplementedError("only support 1-d categoricals")
 
+        assert self.dtype.startswith("int"), self.dtype
+
```
Comment (gfyoung) [inline pandas/io/pytables.py]: Not sure I follow the motivation for this.  Seems out of scope.

Comment (jbrockmendel) [reply pandas/io/pytables.py]: making sure that i understand correctly that the codes is always signed-integer

File: pandas/io/pytables.py [hunk 2]
```diff
@@ -1092,6 +1092,8 @@ def remove(self, key: str, where=None, start=None, stop=None):
         except KeyError:
             # the key is not a valid store, re-raising KeyError
             raise
+        except AssertionError:
```
Comment (jreback) [inline pandas/io/pytables.py]: what are these from?

Comment (jbrockmendel) [reply pandas/io/pytables.py]: these are unrelated to most of the PR, but related to the couple-weeks-ago push to get rid of `except Exception`.  In cases where we can't (yet) get rid of `except Exception`, at least re-raising assertion errors is useful to debugginig

Comment (jbrockmendel) [reply pandas/io/pytables.py]: if this is a sticking point im happy to revert and do this elsewhere

Comment (jreback) [reply pandas/io/pytables.py]: no its fine, if possible can you add a comment indicating where its raised from

**Comment by jreback** [comment]
pls rebase

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by jbrockmendel** [comment]
gentle ping, this is a blocker for the payoff ive been promising
